enhancements/installer/single-node-installation-bootstrap-in-place.md

```diff
@@ -0,0 +1,288 @@
+
+## Motivation
+
+Currently, all OpenShift installations use an auxiliary bootstrap node.
+The bootstrap node creates a temporary control plane that is required for launching the actual cluster.
+
+Resource constraints are the primary motivation for single-node OCP clusters, and removing the need for a
+separate bootstrap machine significantly reduces the resources required to install the cluster.
+
+Requiring the auxiliary node for installing a Single Node OpenShift brings to play some requirements:
+1. The obvious additional node.
+2. Requires external dependencies:
+   a. Load balancer (only for bootstrap phase)
+   b. Requires DNS (configured per installation)
+3. Cannot use Bare Metal IPI:
+   a. Adds irrelevant dependencies - vips, keepalived, mdns
+   b. Requires same L2 between bootstrap and the Single Node OpenShift
+
+### Goals
+
+* This enhancement describes an approach for installing Single Node OpenShift in a BareMetal environment
+  for production use.
+* Minimal changes to openShift installer, this implementation shouldn't effect existing deployment flows.
+* Installation should result a clean Single Node OpenShift, without bootstrap leftovers.
+* Rely on a livecd, so it would be able to trigger the full installation using virtual media.
+* Self-contained bootstrap, The bootstrap flow is expressed as a single Ignition config which is layered on top of a pristine RHCOS image. It's important that we limit our use of this pattern to just the Ignition configs - as opposed to including other assets directly in the ISO image. 
+
+### Non-Goals
+
+* This enhancement does not address similar installation flow for 3 nodes cluster.
+* This enhancement does not address high-availability for single-node deployments.
+* This enhancement does not address single-node-developer cluster-profile installation.
+* This enhancement does not address cloud deployment for bootstrap in place (since livecd cannot be used). It will be addressed as part of future enhancement.
+* This enhancement does not address updating single-node clusters. It will be addressed as part of future enhancement.
+
+## Proposal
+
+Execute the bootstrap flow on a liveCD.
+Enrich the master Ignition with the control plane static pods and write this Ignition.
+Write the updated master Ignition + RHCOS to disk and reboot.
+Post reboot the node will complete the installation.
+
+### User Stories
+
+#### As a user, I can deploy OpenShift in a supported single-node configuration
+
+A user will be able to run the OpenShift installer to create a single-node
+Ignition configuration.
+The user will boot RHCOS liveCD with this Ignition to create a single-node deployment.
+
+### Implementation Details/Notes/Constraints
+
+The OpenShift installer `create ignition-configs` command will generate a `bootstrap-in-place-for-live-iso.ign`
+file, when the number of replicas for the control plane (in the `install-config.yaml`) is `1`.
```

Comment:
I'm assuming the files currently created by `create ignition-configs` when there's only one replica for the control plane will be unchanged, the only difference in this situation before/after this is implemented will be a new `bootstrap-in-place-for-live-iso.ign` file? (in other words, could this impact crc/snc which also run openshift-install with a 1-replica control plane? :)

Reply:
mmm good point. on the other hand, they use libvirt provider. and cluster profile. so maybe we can filter out using those flags

Reply:
According to line 96, new ign file will be generated in parallel, so if you keep on looking into bootstrap.ign you'll not be afffected. If we decide to converge into same bootstrap file, then we'll have to add new ENV OPENSHIFT_INSTALL_EXPERIMENTAL_SINGLE_NODE_LIVE_ISO marking a different path

Reply:
@cfergeau so finally we decided to got with the ENV approach.
https://github.com/openshift/installer/pull/4482/files#diff-cc0fbb14ee7d67263a33bebe0004691706613a326f49183d3748d1cecf46e096R272

Reply:
I'd suggest a different asset entirely (i.e. one asset for bootstrap.ign and another for single-node-bootstrap.ign). That way, a different set of validations can be easily added (e.g. ensure that the number of replicas for the control plane is `1` when creating single-node-bootstrap.ign).